In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
dataset_path = 'World Happiness Report'
dataframe1=pd.read_csv(dataset_path+'/2015.csv')
dataframe2=pd.read_csv(dataset_path+'/2016.csv')
dataframe3=pd.read_csv(dataset_path+'/2017.csv')
dataframe4=pd.read_csv(dataset_path+'/2018.csv')
dataframe5=pd.read_csv(dataset_path+'/2019.csv')
concatenated_dataframes = pd.concat([dataframe1, dataframe2,dataframe3,dataframe4,dataframe5])

In [3]:
concatenated_dataframes.head(10)

Country                     Region  Happiness Rank  Happiness Score  \
0  Switzerland             Western Europe             1.0            7.587   
1      Iceland             Western Europe             2.0            7.561   
2      Denmark             Western Europe             3.0            7.527   
3       Norway             Western Europe             4.0            7.522   
4       Canada              North America             5.0            7.427   
5      Finland             Western Europe             6.0            7.406   
6  Netherlands             Western Europe             7.0            7.378   
7       Sweden             Western Europe             8.0            7.364   
8  New Zealand  Australia and New Zealand             9.0            7.286   
9    Australia  Australia and New Zealand            10.0            7.284   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   
5         0.03140                   1.29025  1.31826   
6         0.02799                   1.32944  1.28017   
7         0.03157                   1.33171  1.28907   
8         0.03371                   1.25018  1.31967   
9         0.04083                   1.33358  1.30923   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  ...  \
0                   0.94143  0.66557                        0.41978  ...   
1                   0.94784  0.62877                        0.14145  ...   
2                   0.87464  0.64938                        0.48357  ...   
3                   0.88521  0.66973                        0.36503  ...   
4                   0.90563  0.63297                        0.32957  ...   
5                   0.88911  0.64169                        0.41372  ...   
6                   0.89284  0.61576                        0.31814  ...   
7                   0.91087  0.65980                        0.43844  ...   
8                   0.90837  0.63938                        0.42922  ...   
9                   0.93156  0.65124                        0.35637  ...   

   Trust..Government.Corruption.  Dystopia.Residual  Overall rank  \
0                            NaN                NaN           NaN   
1                            NaN                NaN           NaN   
2                            NaN                NaN           NaN   
3                            NaN                NaN           NaN   
4                            NaN                NaN           NaN   
5                            NaN                NaN           NaN   
6                            NaN                NaN           NaN   
7                            NaN                NaN           NaN   
8                            NaN                NaN           NaN   
9                            NaN                NaN           NaN   

   Country or region  Score  GDP per capita  Social support  \
0                NaN    NaN             NaN             NaN   
1                NaN    NaN             NaN             NaN   
2                NaN    NaN             NaN             NaN   
3                NaN    NaN             NaN             NaN   
4                NaN    NaN             NaN             NaN   
5                NaN    NaN             NaN             NaN   
6                NaN    NaN             NaN             NaN   
7                NaN    NaN             NaN             NaN   
8                NaN    NaN             NaN             NaN   
9                NaN    NaN             NaN             NaN   

   Healthy life expectancy  Freedom to make life choices  \
0                      NaN                           NaN   
1                      NaN                           NaN   
2                      NaN                           NaN   


In [4]:
concatenated_dataframes.shape

(782, 30)

as we can see above there lots of columns has all null values for first 10 column lets see is it same for all dataframe.


In [5]:
concatenated_dataframes.isnull().sum()

Country                          312
Region                           467
Happiness Rank                   467
Happiness Score                  467
Standard Error                   624
Economy (GDP per Capita)         467
Family                           312
Health (Life Expectancy)         467
Freedom                          312
Trust (Government Corruption)    467
Generosity                         0
Dystopia Residual                467
Lower Confidence Interval        625
Upper Confidence Interval        625
Happiness.Rank                   627
Happiness.Score                  627
Whisker.high                     627
Whisker.low                      627
Economy..GDP.per.Capita.         627
Health..Life.Expectancy.         627
Trust..Government.Corruption.    627
Dystopia.Residual                627
Overall rank                     470
Country or region                470
Score                            470
GDP per capita                   470
Social support                   470
H

we have 782 rows and there is no all nan values column.

In [6]:
concatenated_dataframes.dtypes

Country                           object
Region                            object
Happiness Rank                   float64
Happiness Score                  float64
Standard Error                   float64
Economy (GDP per Capita)         float64
Family                           float64
Health (Life Expectancy)         float64
Freedom                          float64
Trust (Government Corruption)    float64
Generosity                       float64
Dystopia Residual                float64
Lower Confidence Interval        float64
Upper Confidence Interval        float64
Happiness.Rank                   float64
Happiness.Score                  float64
Whisker.high                     float64
Whisker.low                      float64
Economy..GDP.per.Capita.         float64
Health..Life.Expectancy.         float64
Trust..Government.Corruption.    float64
Dystopia.Residual                float64
Overall rank                     float64
Country or region                 object
Score           

Table has lots of nan values if we fill nan values with mean or median because it will manipulate all dataset. Also, if we remove rows which has nan values, we will lose major of our dataset or may be we will lose all of them.

In [31]:
# Question 1
# Which countries or regions rank highest in overall happiness and each of the six factors that contribute to happiness?



# Grouping by country and getting mean value
mean_values = concatenated_dataframes.groupby('Country').mean()

# Find the top 10 countries with highest overall happiness scores
top_10_countries = mean_values.sort_values("Happiness.Score", ascending=False).head(10)
print("Top 10 countries with highest overall happiness scores:\n")
print(top_10_countries[["Happiness.Score"]])

# Find the top 10 countries with highest scores for each of the six factors

print("\nTop 10 countries with highest scores for each of the six factors:\n")

for factor in mean_values.columns[[3,5,6,7,8,9]]:
    top_10_countries = mean_values.sort_values(factor, ascending=False).head(10)
    print(f"{factor}:\n")
    print(top_10_countries[[factor]])
    print()

Top 10 countries with highest overall happiness scores:

             Happiness.Score
Country                     
Norway                 7.537
Denmark                7.522
Iceland                7.504
Switzerland            7.494
Finland                7.469
Netherlands            7.377
Canada                 7.316
New Zealand            7.314
Australia              7.284
Sweden                 7.284

Top 10 countries with highest scores for each of the six factors:

Economy (GDP per Capita):

                      Economy (GDP per Capita)
Country                                       
Qatar                                 1.757345
Luxembourg                            1.630715
Kuwait                                1.585680
Singapore                             1.583705
Norway                                1.518220
United Arab Emirates                  1.500395
Switzerland                           1.461920
United States                         1.451235
Hong Kong                     

In [47]:
# Question 2
# How did country rankings or scores change between 2015-2016 and 2016-2017 reports?

# Find the change in overall happiness scores between 2015-2016 and 2016-2017


#Create a new column in each dataset to indicate the year
dataframe1['Year'] = 2015
dataframe2['Year'] = 2016
dataframe3['Year'] = 2017

# in dataframe 3 Happiness Score column is named as Happines.Score we should change name for correct merge process
dataframe3 = dataframe3.rename(columns={'Happiness.Score': 'Happiness Score'})


# Merge the two datasets on the 'Country' column, which contains the country names
merged_data15_16 = pd.merge(dataframe1, dataframe2, on='Country')
merged_data16_17 = pd.merge(dataframe2, dataframe3, on='Country')

# Calculate the difference in happiness score between 2015-2016 and 2016-2017 for each country
merged_data15_16['Score Difference'] = merged_data15_16['Happiness Score_y'] - merged_data15_16['Happiness Score_x']
merged_data16_17['Score Difference'] = merged_data16_17['Happiness Score_y'] - merged_data16_17['Happiness Score_x']









                Country  Score Difference
66              Algeria             0.750
150                Togo             0.464
87               Latvia             0.462
84              Romania             0.404
98              Hungary             0.345
135             Senegal             0.315
101             Tunisia             0.306
97              Lebanon             0.290
115               Nepal             0.279
93   Dominican Republic             0.270
       Country  Score Difference
110    Liberia            -0.949
21   Venezuela            -0.726
113      Haiti            -0.490
109   Zimbabwe            -0.417
112      Sudan            -0.411
13      Mexico            -0.409
76     Nigeria            -0.393
121   Botswana            -0.358
105    Ukraine            -0.357
108      Ghana            -0.357
          Country  Score Difference
122      Bulgaria             0.497
146   Afghanistan             0.434
148         Syria             0.393
113         Egypt             0

In [48]:
# Print the countries that experienced the most significant increase or decrease in happiness score
print('Happiness score differences between 2015 and 2016\n')
print(merged_data15_16.sort_values(by='Score Difference', ascending=False).head(10)[['Country','Score Difference']])
print()
print(merged_data15_16.sort_values(by='Score Difference', ascending=True).head(10)[['Country','Score Difference']])


Happiness score differences between 2015 and 2016

                Country  Score Difference
66              Algeria             0.750
150                Togo             0.464
87               Latvia             0.462
84              Romania             0.404
98              Hungary             0.345
135             Senegal             0.315
101             Tunisia             0.306
97              Lebanon             0.290
115               Nepal             0.279
93   Dominican Republic             0.270

       Country  Score Difference
110    Liberia            -0.949
21   Venezuela            -0.726
113      Haiti            -0.490
109   Zimbabwe            -0.417
112      Sudan            -0.411
13      Mexico            -0.409
76     Nigeria            -0.393
121   Botswana            -0.358
105    Ukraine            -0.357
108      Ghana            -0.357


In [50]:
print('Happiness score differences between 2016 and 2017\n')
print(merged_data16_17.sort_values(by='Score Difference', ascending=False).head(10)[['Country','Score Difference']])
print()
print(merged_data16_17.sort_values(by='Score Difference', ascending=True).head(10)[['Country','Score Difference']])

Happiness score differences between 2016 and 2017

          Country  Score Difference
122      Bulgaria             0.497
146   Afghanistan             0.434
148         Syria             0.393
113         Egypt             0.373
109  South Africa             0.370
127         Gabon             0.344
138        Uganda             0.342
121       Senegal             0.316
97       Honduras             0.310
67        Romania             0.297

         Country  Score Difference
40     Venezuela            -0.834
35       Algeria            -0.483
129        Haiti            -0.425
124     Zimbabwe            -0.318
141     Tanzania            -0.317
15        Brazil            -0.317
71       Somalia            -0.289
99        Zambia            -0.281
23        Panama            -0.249
135  South Sudan            -0.241


In [67]:
# Question 3
# Has any country experienced a significant increase or decrease in happiness?

# calculate the difference in happiness scores between consecutive years
concatenated_dataframes["happiness_change"] = concatenated_dataframes.groupby("Country")["Happiness Score"].shift(periods=-1) - concatenated_dataframes["Happiness Score"]

# create a new column that indicates whether the change in happiness score is significant
concatenated_dataframes["change_significant"] = concatenated_dataframes["happiness_change"].apply(lambda x: 1 if x > 0.5 or x < -0.5 else 0)

# group by country and calculate the proportion of years in which the change in happiness score is significant
significant_change = concatenated_dataframes.groupby("Country")["change_significant"].mean()

# display the countries that experienced a significant increase or decrease in happiness
print(significant_change[significant_change > 0].index)

Index(['Algeria', 'Liberia', 'Venezuela'], dtype='object', name='Country')
